.eleventy.js

```diff
@@ -67,6 +68,10 @@ module.exports = function (config) {
         return getResourceLinks(content, sideContentHeadings, lang);
     });
 
+    config.addShortcode("article_contents", (content, headingsSelector) => {
```

Comment:
I'm starting to think this may make more sense as as [filter](https://www.11ty.dev/docs/filters/) instead of a shortcode. In that way you'd have more control of where it is placed and can also run it on different parts of the content if you'd like. You'd also just return the markup of the Table of Contents.

for example in the topic-page template you might have something like:

```nunjucks
{% block content %}
     {{ content | article_contents(translations[lang].articleContentsSummary) }}
     {{ content }}
{% endblock %}
```

or if you wanted it in an aside:

```nunjucks
{% block content %}
     <aside> {{ content | article_contents(translations[lang].articleContentsSummary) }} </aside>
     {{ content }}
{% endblock %}
```

Reply:
This is an interesting question, and one I pondered a little when I made it a shortcode in the first place. I vaguely recall trying it out as a filter before changing tack, though I haven't found any notes I wrote during that decision process. I do see how the idea of a filter makes sense in that the content "passes through" this function before being rendered to the page. I do also like the idea of it appearing separately from the content in the template markup. This is as I'd originally envisioned before getting stuck trying to make a shortcode that returned a wrapped object to unpack and use in the template; @cindyli reminded me that they can only return strings and template literals, and I became unstuck and implemented what you see now.

Reviewing the 11ty docs pages for [filters](https://www.11ty.dev/docs/filters/) and [shortcodes](https://www.11ty.dev/docs/shortcodes/) (which are a bit sparse about their intended uses), I'm still a bit unsure but am currently leaning towards sticking with shortcodes. On one hand, the filters doc says that they "transform data into a more presentable format", which it could easily be argued this function does by adding this new navigational structure/markup. On the other hand, the definition provided in the shortcodes doc neatly describes what `article_contents` currently does:
> A shortcode returns content (a JavaScript string or template literal) that is injected into the template. You can use these however you’d like—you could even think of them as reusable components.

What do you think?

Reply:
In short I'm less concerned about the actual mechanism but more concerned with the ability to control where things go. In thinking about it more, you could achieve what I suggested above using a shortcode, by not appending the content. So I'd probably still argue for outputting only the table of contents, but if you feel a shortcode is more appropriate no problem.

I know for the time being it will seem more efficient to just return everything in the shortcode. At least for now the output will be identical. However, I feel this will give you more freedom to make changes, or resume the `article_contents` in the future. Also, if you were able to use Github_slugger and remove the call to fluid.allocateSimpleId, there is no need to operate on the rest of the content, so it reduces changes for any side affects or confusion about what is actually happening.